### PR TITLE
523: Add JSON formats

### LINKED ISSUE #523: Add JSON format
For WordPress core we are [looking into](https://core.trac.wordpress.org/ticket/20491) adding some more JavaScript i18n tools so we can more easily translate strings in JavaScript. Instead of loading `.mo` files and passing only few of those strings to the JS, it would make more sense to have all JS-related strings in a single `.json` file.

For this, GlotPress needs a new format class to enable exporting JSON files. Ideally, these would be compatible with the format [Jed](https://github.com/SlexAxton/Jed) expects, but I can also see how this might not be ideal. If that's the case, translate.wordpress.org could always override/extend any JSON format in GlotPress.

Looking at #297 and existing formats, implementing this should be quite easy and I would be more than happy to work on a patch for this.

### PR DESCRIPTION
**Pull request by swissspidy**
This integrates the [glotpress-json](https://github.com/swissspidy/glotpress-json) plugin I've been working on into GlotPress.

Adds two new formats (including tests):
- `GP_Format_JSON`: "simple" JSON (key -> value)
- `GP_Format_Jed1x`: Jed 1.x compatible format, extends `GP_Format_JSON`

As you can see, the `GP_Format_JSON` class is easily extensible for more specific use cases. See swissspidy/glotpress-json#4 for some background.

Fixes #523.

### CODE REVIEW

**Comment by swissspidy** [comment]
Note: Because both formats use the `.json` file extension, `test_gp_get_import_file_format()` currently fails. Any ideas welcome.

**Comment by ocean90** [comment]
>  `test_gp_get_import_file_format()` currently fails. 

Maybe `gp_get_import_file_format()` needs to be changed to return an array of formats if multiple formats exist for a file extension. In `GP_Route_Translation::import_translations_post()` the import should then be canceled if the extension is ambiguous. @toolstack What do you think?

**Comment by toolstack** [comment]
> > test_gp_get_import_file_format() currently fails. 
> 
> Maybe gp_get_import_file_format() needs to be changed to return an array of formats if multiple formats exist for a file extension. In GP_Route_Translation::import_translations_post() the import should then be canceled if the extension is ambiguous. @toolstack What do you think?

I don't like the idea of returning an array as it could break many perfectly useable cases.  We already have this in the Android/ResX formats, both use .xml as their file extension, but we use .resx.xml for ResX.

I'd change the extension name for the JED files to something like ".jed1x.json" so they can be identified uniquely.  Or if we think JED is the more widely used format, change the base class to ".raw.json" or something similar.

**Comment by ocean90** [comment]
@toolstack Interesting, I wasn't aware of that. Isn't .resx also an extension for the ResX format? Should we update the name property for the dropdown?

format-dropdown

**Comment by toolstack** [comment]
@ocean90 Yea, I found out about it when I was doing the auto detection code.  Looks like .resx is a valid extension so we should update it, I'll create a PR once we merge the file rename PR.

**Comment by swissspidy** [comment]
OK, so changing the extensions sounds like an easy solution. I'd go for `.jed.json` or `.jed1x.json` in that case

**Comment by swissspidy** [comment]
Just to let you know, this PR now adds meta data to the Jed 1.x format as discussed in swissspidy/glotpress-json#4.

File: gp-includes/formats/format-jed1x.php [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+		return $entries;
+	}
+
+	/**
+	 * Decodes a JSON string and checks for needed array keys.
+	 *
+	 * @since 2.3.0
+	 *
+	 * @param string $file_name The name of the JSON file to parse.
+	 * @return array|false The encoded value or fals on failure.
```
Comment (ocean90) [inline gp-includes/formats/format-jed1x.php]: Typo: `fals` => `false`

File: gp-includes/formats/format-json.php [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+	 *
+	 * @param GP_Project         $project         The project the strings are being exported for, not used
+	 *                                            in this format but part of the scaffold of the parent object.
+	 * @param GP_Locale          $locale          The locale object the strings are being exported for. not used
+	 *                                            in this format but part of the scaffold of the parent object.
+	 * @param GP_Translation_Set $translation_set The locale object the strings are being
+	 *                                            exported for. not used in this format but part
+	 *                                            of the scaffold of the parent object.
+	 * @param GP_Translation     $entries         The entries to export.
+	 * @return string
```
Comment (ocean90) [inline gp-includes/formats/format-json.php]: Missing description.

File: gp-includes/formats/format-json.php [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+		return wp_json_encode( $result, ( true === $pretty_print ) ? JSON_PRETTY_PRINT : 0 );
+	}
+
+	/**
+	 * Reads a set of original strings from a JSON file.
+	 *
+	 * @since 2.3.0
+	 *
+	 * @param string $file_name The name of the uploaded JSON file.
+	 * @return Translations|bool
```
Comment (ocean90) [inline gp-includes/formats/format-json.php]: Missing description.

File: gp-includes/formats/format-json.php [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+	}
+
+	/**
+	 * Reads a set of translations from a JSON file.
+	 *
+	 * @since 2.3.0
+	 *
+	 * @param string     $file_name The name of the uploaded properties file.
+	 * @param GP_Project $project   Unused. The project object to read the translations into.
+	 * @return Translations
```
Comment (ocean90) [inline gp-includes/formats/format-json.php]: Missing description.

File: gp-includes/formats/format-json.php [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+		return $this->read_originals_from_file( $file_name );
+	}
+
+	/**
+	 * Decodes a JSON string.
+	 *
+	 * @since 2.3.0
+	 *
+	 * @param string $file_name The name of the JSON file to parse.
+	 * @return array|false The encoded value or fals on failure.
```
Comment (ocean90) [inline gp-includes/formats/format-json.php]: Typo: `fals` => `false`

File: gp-includes/formats/format-jed1x.php [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+		return str_replace( '__GP_EMPTY__', '', $result );
+	}
+
+	/**
+	 * Reads a set of original strings from a JSON file.
+	 *
+	 * @since 2.3.0
+	 *
+	 * @param string $file_name The name of the uploaded JSON file.
+	 * @return Translations|bool
```
Comment (ocean90) [inline gp-includes/formats/format-jed1x.php]: Missing description.

File: gp-includes/formats/format-jed1x.php [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+		$result = wp_json_encode( $result, ( true === $pretty_print ) ? JSON_PRETTY_PRINT : 0 );
+
+		/*
+		 * Replace '__GP_EMPTY__' with an acttual empty string.
+		 *
+		 * Empty object property names are not supported in PHP, so they would get lost.
+		 *
+		 * Note: When decoding, PHP replaces empty strings with '_empty_'.
+		 *
+		 * @see https://bugs.php.net/bug.php?id=50867
```
Comment (ocean90) [inline gp-includes/formats/format-jed1x.php]: `@see` should be `@link`

File: gp-includes/formats/format-jed1x.php [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+			} );
+		}
+
+		/** This filter is documented in gp-includes/formats/format-json.php */
+		$pretty_print = apply_filters( 'gp_json_export_pretty_print', false );
+
+		$result = wp_json_encode( $result, ( true === $pretty_print ) ? JSON_PRETTY_PRINT : 0 );
+
+		/*
+		 * Replace '__GP_EMPTY__' with an acttual empty string.
```
Comment (ocean90) [inline gp-includes/formats/format-jed1x.php]: Typo: `acttual` => `actual`

File: gp-includes/formats/format-jed1x.php [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+	 *
+	 * @param GP_Project         $project         The project the strings are being exported for, not used
+	 *                                            in this format but part of the scaffold of the parent object.
+	 * @param GP_Locale          $locale          The locale object the strings are being exported for. not used
+	 *                                            in this format but part of the scaffold of the parent object.
+	 * @param GP_Translation_Set $translation_set The locale object the strings are being
+	 *                                            exported for. not used in this format but part
+	 *                                            of the scaffold of the parent object.
+	 * @param GP_Translation     $entries         The entries to export.
+	 * @return string
```
Comment (ocean90) [inline gp-includes/formats/format-jed1x.php]: Missing description.

File: gp-includes/formats/format-jed1x.php [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+	public $extension = 'jed.json';
+
+	/**
+	 * Generates a string the contains the $entries to export in the JSON file format.
+	 *
+	 * @since 2.3.0
+	 *
+	 * @param GP_Project         $project         The project the strings are being exported for, not used
+	 *                                            in this format but part of the scaffold of the parent object.
+	 * @param GP_Locale          $locale          The locale object the strings are being exported for. not used
```
Comment (ocean90) [inline gp-includes/formats/format-jed1x.php]: Typo: `for. not used` => `for, not used` (comma)

File: gp-includes/formats/format-json.php [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+	 * @param string     $file_name The name of the uploaded properties file.
+	 * @param GP_Project $project   Unused. The project object to read the translations into.
+	 * @return Translations
+	 */
+	public function read_translations_from_file( $file_name, $project = null ) {
+		return $this->read_originals_from_file( $file_name );
+	}
+
+	/**
+	 * Decodes a JSON string.
```
Comment (ocean90) [inline gp-includes/formats/format-json.php]: I think the summary should mention that the methods reads the content of a file and then decodes the content. Maybe the name of the method should reflect that too: `decode_json_file( $file_name)`

**Review by ocean90 — CHANGES_REQUESTED** [review]
There are some missing `@return` descriptions and I suggest to rename `GP_Format_JSON::json_encode()`.

**Comment by swissspidy** [comment]
@ocean90 Thanks for the review. I've addressed these points now.

**Review by ocean90 — APPROVED** [review]
LGTM.

**Pull request merged**